ml/train/losses/losses.go

```diff
@@ -398,3 +398,211 @@ func MakeHuberLoss(delta float64) LossFn {
+func maskedMaximum(x, mask *Node, dim int) *Node {
+	axisMinimums := ReduceAndKeep(x, ReduceMin, dim)
+	return Add(
+		ReduceAndKeep(
+			Mul(
+				Sub(x, axisMinimums),
+				ConvertDType(mask, x.DType())),
+			ReduceMax,
+			dim),
+		axisMinimums)
+}
+
+//  Computes the axis wise minimum over chosen elements.
+
+//     Args:
+//       data: 2-D float `Tensor` of shape `[n, m]`.
+//       mask: 2-D Boolean `Tensor` of shape `[n, m]`.
+//       dim: The dimension over which to compute the minimum.
+
+// Returns:
+//
+//	masked_minimums: N-D `Tensor`.
+//	  The minimized dimension is of size 1 after the operation.
+func maskedMinimum(x, mask *Node, dim int) *Node {
+	axisMaximums := ReduceAndKeep(x, ReduceMax, dim)
+	return Add(
+		ReduceAndKeep(
+			Mul(
+				Sub(x, axisMaximums),
+				ConvertDType(mask, x.DType())),
+			ReduceMin,
+			dim,
+		),
+		axisMaximums)
+}
+
+// Computes the triplet loss with semi-hard negative mining
+//
+//		labels: 1-D integer `Tensor` with shape `[batch_size]` of multiclass integer labels.
+//		predictions: 2-D float `Tensor` of embedding vectors. Embeddings should be l2 normalized.
+//	    margin: Float, margin term in the loss definition.
+//	    distance_metric: `str` that determines distance metric.
+//	        Valid strings are "L2" for l2-norm distance,
+//	        "squared-L2" for squared l2-norm distance,
+//	        and "angular" for cosine similarity.
+func TripletSemiHardLoss(labels, predictions []*Node, margin float64, distance string) *Node {
+	predictions0 := predictions[0]
+	labels0 := labels[0]
+	if labels0.Shape().Rank() != predictions0.Shape().Rank() {
+		Panicf("labels[0] (%s) and predictions[0] (%s) must have the same rank", labels0.Shape(), predictions0.Shape())
+	}
+	weights, mask := CheckLabelsForWeightsAndMask(labels0.Shape(), labels)
+
+	// Build pairwise square distance matrix
+	var pairwiseDist *Node
+	switch distance {
+	case "L2":
+		pairwiseDist = pairwiseDistance(predictions0, false)
+	case "squared-L2":
+		pairwiseDist = pairwiseDistance(predictions0, true)
+	case "angular":
+		angularDistance(predictions0)
+	default:
+		Panicf("distance %s don't exist or not implemented yet", distance)
+	}
+
+	// Build pairwise binary adjacency matrix.
+	adjacency := Equal(labels0, Transpose(labels0, 0, 1))
+	// Invert so we can select negatives only.
+	adjacencyNot := LogicalNot(adjacency)
+
+	batchSize := labels0.Shape().Size()
+
+	// Compute the mask.
+	pairwiseDistRepeated := Reshape(
+		BroadcastToDims(
+			ExpandAxes(pairwiseDist, 0),
+			batchSize, batchSize, batchSize),
+		batchSize*batchSize, batchSize)
+	pairwiseDistMask := And(
+		Reshape(
+			BroadcastToDims(
+				ExpandAxes(adjacencyNot, 0),
+				batchSize, batchSize, batchSize),
+			batchSize*batchSize, batchSize),
+		GreaterThan(
+			pairwiseDistRepeated,
+			Reshape(Transpose(pairwiseDist, 0, 1), -1, 1),
+		),
+	)
+	pairwiseDistMaskFinal := Transpose(
+		Reshape(
+			GreaterThan(
+				ReduceAndKeep(pairwiseDistMask, ReduceSum, 1),
```

Comment:
Since `pairwiseDistMask` is boolean, what about using `LogicalAny()` here ? (It's like a `ReduceOr` if you will).

This is because `ReduceSum` of boolean values can have an ambiguous interpretation. What is `true+true+false+false+false` ?  For that reason there is `LogicalAny()` and `LogicalAnd`. They also should work with `ReduceAndKeep`.